plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,186 @@
+            }
+        });
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Map<String, QueryGroup> existingQueryGroups = currentClusterState.metadata().queryGroups();
+        String groupName = queryGroup.getName();
+
+        // check if maxQueryGroupCount will breach
+        if (existingQueryGroups.size() + 1 > maxQueryGroupCount) {
+            logger.warn("{} value exceeded its assigned limit of {}", QUERY_GROUP_COUNT_SETTING_NAME, maxQueryGroupCount);
+            throw new RuntimeException("Can't create more than " + maxQueryGroupCount + " QueryGroups in the system");
+        }
+
+        // check for duplicate name
+        Optional<QueryGroup> findExistingGroup = existingQueryGroups.values()
+            .stream()
+            .filter(group -> group.getName().equals(groupName))
+            .findFirst();
+        if (findExistingGroup.isPresent()) {
+            logger.warn("QueryGroup with name {} already exists. Not creating a new one.", groupName);
+            throw new RuntimeException("QueryGroup with name " + groupName + " already exists. Not creating a new one.");
+        }
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        Map<ResourceType, Double> existingUsageMap = calculateExistingUsage(existingQueryGroups, groupName);
+        for (ResourceType resourceType : queryGroup.getResourceLimits().keySet()) {
+            double totalUsage = existingUsageMap.getOrDefault(resourceType, 0.0) + queryGroup.getResourceLimits().get(resourceType);
+            if (totalUsage > 1) {
+                logger.warn("Total resource allocation for {} will go above the max limit of 1.0", resourceType.getName());
+                throw new RuntimeException(
+                    "Total resource allocation for " + resourceType.getName() + " will go above the max limit of 1.0"
+                );
+            }
+        }
+
+        return ClusterState.builder(currentClusterState)
+            .metadata(Metadata.builder(currentClusterState.metadata()).put(queryGroup).build())
+            .build();
+    }
+
+    /**
+     * This method calculates the existing total usage of the all the resource limits (except the group that we're updating here)
+     * @param existingQueryGroups - existing QueryGroups
+     * @param groupName - the QueryGroup name we're updating
+     */
+    private Map<ResourceType, Double> calculateExistingUsage(Map<String, QueryGroup> existingQueryGroups, String groupName) {
+        Map<ResourceType, Double> map = new HashMap<>();
+        for (QueryGroup currGroup : existingQueryGroups.values()) {
+            if (!currGroup.getName().equals(groupName)) {
```

Comment:
Although not required for create, this might be need for update in future. So, we can keep it as is